src/linalg/Svd_truncate.cpp

```diff
@@ -12,53 +13,77 @@ namespace cytnx {
-      std::vector<Tensor> tmps = Svd(Tin, is_UvT);
+      if (Tin.device() == Device.cpu) {
+        std::vector<Tensor> tmps = Svd(Tin, is_UvT);
 
-      cytnx_uint64 id = 0;
-      cytnx_uint64 Kdim = keepdim;
+        cytnx_uint64 id = 0;
+        cytnx_uint64 Kdim = keepdim;
 
-      Storage ts = tmps[0].storage();
+        Storage ts = tmps[0].storage();
 
-      if (ts.size() < keepdim) {
-        Kdim = ts.size();
-      }
+        if (ts.size() < keepdim) {
+          Kdim = ts.size();
+        }
 
-      cytnx_uint64 truc_dim = Kdim;
-      for (cytnx_int64 i = Kdim - 1; i >= 0; i--) {
-        if (ts.at(i) < err) {
-          truc_dim--;
-        } else {
-          break;
+        cytnx_uint64 truc_dim = Kdim;
+        for (cytnx_int64 i = Kdim - 1; i >= 0; i--) {
+          if (ts.at(i) < err) {
+            truc_dim--;
+          } else {
+            break;
+          }
         }
-      }
 
-      if (truc_dim == 0) {
-        truc_dim = 1;
-      }
-      /// std::cout << truc_dim << std::endl;
-      // cytnx_error_msg(tmps[0].shape()[0] < keepdim,"[ERROR] keepdim should be <= the valid # of
-      // singular value, %d!\n",tmps[0].shape()[0]);
-      Tensor terr({1}, Type.Double);
+        if (truc_dim == 0) {
+          truc_dim = 1;
+        }
+        /// std::cout << truc_dim << std::endl;
+        // cytnx_error_msg(tmps[0].shape()[0] < keepdim,"[ERROR] keepdim should be <= the valid # of
+        // singular value, %d!\n",tmps[0].shape()[0]);
+        Tensor terr({1}, Type.Double);
 
-      if (truc_dim != ts.size()) {
-        if (return_err == 1)
-          terr = tmps[id](truc_dim);
-        else if (return_err)
-          terr = tmps[id].get({ac::tilend(truc_dim)});
+        if (truc_dim != ts.size()) {
+          if (return_err == 1)
+            terr = tmps[id](truc_dim);
+          else if (return_err)
+            terr = tmps[id].get({ac::tilend(truc_dim)});
 
-        tmps[id] = tmps[id].get({ac::range(0, truc_dim)});
+          tmps[id] = tmps[id].get({ac::range(0, truc_dim)});
 
-        if (is_UvT) {
-          id++;
-          tmps[id] = tmps[id].get({ac::all(), ac::range(0, truc_dim)});
+          if (is_UvT) {
+            id++;
+            tmps[id] = tmps[id].get({ac::all(), ac::range(0, truc_dim)});
 
-          id++;
-          tmps[id] = tmps[id].get({ac::range(0, truc_dim), ac::all()});
+            id++;
+            tmps[id] = tmps[id].get({ac::range(0, truc_dim), ac::all()});
+          }
         }
-      }
-      if (return_err) tmps.push_back(terr);
+        if (return_err) tmps.push_back(terr);
+
+        return tmps;
+      } else {
+#ifdef UNI_GPU
+        std::vector<Tensor> tmps = Svd(Tin, is_UvT);
+        Tensor terr({1}, Tin.dtype(), Tin.device());
 
-      return tmps;
+        cytnx::linalg_internal::lii.cudaMemcpyTruncation_ii[Tin.dtype()](
+          tmps[1], tmps[2], tmps[0], terr, keepdim, err, is_UvT, is_UvT, return_err);
```

Comment:
Cool! should we do the same for CPU side ?

Reply:
Yes, I will add the similar functions for cpu case.